text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,231 @@
+- Feature Name: `packages_as_namespaces`
+- Start Date: (fill me in with today's date, 2022-03-09)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Languages like C++ have open namespaces where anyone can write code in any namespace.  In C++'s case, this includes the `std` namespace and is only limited by convention.  In contrast, Rust has closed namespaces which can only include code from the original namespace definition (the crate).
```

Comment:
This is a pedantic nit not particularly relevant to Rust, but C++ declares the declaration of names in the `std` namespace to be **UB** (or at least **Ill-formed, no diagnostic required**; I couldn't find a normative reference offhand), so it's not *just* convention that limits the `std` namespace.

Every other namespace *is* only managed by convention in C++, though.

Reply:
Maybe I'm being blasé with the standard but that sounds like "by convention" to me in that there is no enforcement of that at compile time but instead a document just says "don't do it"

Reply:
As I said, it's a pedantic nit that doesn't matter much, but I do think there's a difference between convention and mandate (even unenforced). You can turn off Rust's naming conventions and suffer nothing but funny looks, but if you violate the C++ mandate that user code shall not define names in the std namespace you're actively courting disaster.

But I don't quite know where the cutoff point is. E.g. saying that safe Rust code is only safe by convention sounds quite wrong, but nothing is *factually incorrect* about that statement; the Rust compiler won't prevent you from writing a safe function that just calls an `unsafe` one if you've decided to do so.